packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -0,0 +1,204 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import type { Type } from 'typescript';
+
+import * as util from '../util';
+
+export type Options = [
+  {
+    ignoreIntersections?: boolean;
+    ignoreUnions?: boolean;
+  },
+];
+
+export type MessageIds = 'duplicate';
+
+const isRecordType = (
+  object: object,
+): object is Record<string | symbol, unknown> => {
+  return object.constructor === Object;
+};
+
+const astIgnoreKeys = ['range', 'loc', 'parent'];
+
+const isSameAstNode = (actualNode: unknown, expectedNode: unknown): boolean => {
+  if (actualNode === expectedNode) {
+    return true;
+  }
+  if (
+    actualNode &&
+    expectedNode &&
+    typeof actualNode == 'object' &&
+    typeof expectedNode == 'object'
+  ) {
+    if (actualNode.constructor !== expectedNode.constructor) {
+      return false;
+    }
+    if (Array.isArray(actualNode) && Array.isArray(expectedNode)) {
+      if (actualNode.length != expectedNode.length) {
+        return false;
+      }
+      return !actualNode.some(
+        (nodeEle, index) => !isSameAstNode(nodeEle, expectedNode[index]),
+      );
+    }
+    if (!isRecordType(actualNode) || !isRecordType(expectedNode)) {
+      return false;
+    }
```

Comment:
This line looks like it's just here for type system coverage. I'm guessing you'll have to use some `as` assertions later on instead of it.